lib/portage/process.py

```diff
@@ -232,12 +239,16 @@ async def run_coroutine_exitfuncs():
     This is the same as run_exitfuncs but it uses asyncio.iscoroutinefunction
     to check which functions to run. It is called by the AsyncioEventLoop
     _close_main method just before the loop is closed.
+
+    If the loop is explicitly closed before exit, then that will cause
+    run_coroutine_exitfuncs to run before run_exitfuncs. Otherwise, a
+    run_exitfuncs hook will close it, causing run_coroutine_exitfuncs to be
+    called via run_exitfuncs.
     """
     tasks = []
-    for index, (func, targs, kargs) in reversed(list(enumerate(_exithandlers))):
-        if asyncio.iscoroutinefunction(func):
-            del _exithandlers[index]
-            tasks.append(asyncio.ensure_future(func(*targs, **kargs)))
+    while _coroutine_exithandlers:
+        func, targs, kargs = _coroutine_exithandlers.pop()
+        tasks.append(asyncio.ensure_future(func(*targs, **kargs)))
```

Comment:
I've created a test case that starts a socks5 proxy from a subprocess and then calls `run_exitfuncs()` to test the cleanup. This has revealed a deadlock in `_thread_weakrefs_atexit()` that was resolved by making the lock reentrant.

Reply:
I changed it to release the lock while closing the loop, and then the reentrant lock was not needed to avoid a deadlock.